File: concepts/graphs/directed/topological_sort.py
from collections import defaultdict

class Solution:
    def __init__(self) -> None:
        self.visited = defaultdict(bool)
        self.stack = []
    
    def dfs(self, adj, curr):
        
        self.visited[curr] = True

        for child in adj[curr]:
            if not self.visited[child]:
                self.dfs(adj=adj, curr=child)
            
        self.stack.append(curr)


    #Function to return list containing vertices in Topological order.
    def topoSort(self, V, adj):
        for i in range(0, len(adj)):
            if (not self.visited[i]):
                self.dfs(adj, i)

        result = []
        while(len(self.stack) > 0):
            peek = self.stack[-1]
            result.append(peek)
            self.stack.pop(-1)

        return result







ans = Solution().topoSort(V=4, adj=[[0], [], [0], [0]])
print(ans)

File: concepts/graphs/directed/topological_sort.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
   public:
    void dfs(vector<int> adj[], int curr, vector<int> &visited, stack<int> &st) {
        visited[curr] = true;
        for (int &child : adj[curr]) {
            if (!visited[curr]) {
                dfs(adj, child, visited, st);
            }
        }
        st.push(curr);
    }

    vector<int> topoSort(int V, vector<int> adj[]) {
        vector<int> visited(V, false);
        stack<int> st;

        for (int i = 0; i < V; i++) {
            if (!visited[i]) {
                dfs(adj, i, visited, st)
            }
        }

        vector<int> result;

        while (!st.empty()) {
            int temp = st.top();
            result.push_back(temp);
            st.pop();
        }

        return result;
    }
};
